cmd/snapd-generator/main.c

```diff
@@ -30,30 +31,168 @@
 #include "../libsnap-confine-private/mountinfo.h"
 #include "../libsnap-confine-private/string-utils.h"
 
-static sc_mountinfo_entry *find_root_mountinfo(sc_mountinfo * mounts)
+static sc_mountinfo_entry *find_dir_mountinfo(const char *mnt_dir)
 {
+	// Load /proc/1/mountinfo so that we can inspect the root filesystem.
+	sc_mountinfo *mounts SC_CLEANUP(sc_cleanup_mountinfo) = NULL;
+	mounts = sc_parse_mountinfo("/proc/1/mountinfo");
+	if (!mounts) {
+		fprintf(stderr, "cannot open or parse /proc/1/mountinfo\n");
+		return NULL;
+	}
+
 	sc_mountinfo_entry *cur, *root = NULL;
 	for (cur = sc_first_mountinfo_entry(mounts); cur != NULL;
 	     cur = sc_next_mountinfo_entry(cur)) {
 		// Look for the mount info entry for the root file-system.
-		if (sc_streq("/", cur->mount_dir)) {
+		if (sc_streq(mnt_dir, cur->mount_dir)) {
 			root = cur;
+			break;
```

Comment:
Technically we should not break, this is because the list can be over-mounted and many things may effectively be mounted in the same place. The last one is the one that wins.

Reply:
Ok, I've added a comment about this